tutorials/37-offline.md

```diff
@@ -6,30 +6,34 @@ Enketo may be [configured](./tutorial-10-configure.html#offline-enabled) to supp
 
 #### Service Worker and Cache Storage
 
-The Enketo app registers a [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API), which intercepts some network requests in order to [cache](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage) certain resources as they're requested. These resources include:
+The Enketo app registers a [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API), which intercepts some network requests in order to [cache](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage) certain resources as they're requested. The following resources are stored in one of two caches:
 
--   The Service Worker script itself
--   The Enketo app's static assests (core JavaScript and CSS)
--   CSS supplied by [configured themes](./tutorial-10-configure.html#themes-supported)
--   The Enketo app icon, which may be displayed if a user bookmarks the app or registers it as a mobile web app
--   The initial HTML which begins loading the app and the requested form
+-   Common static assets used by Enketo to serve and render forms:
 
-This cache is cleared and rebuilt, after a brief delay, when:
+    -   The Service Worker script itself
+    -   The Enketo app's static assests (core JavaScript and CSS)
+    -   CSS supplied by [configured themes](./tutorial-10-configure.html#themes-supported)
+    -   The Enketo app icon, which may be displayed if a user bookmarks the app or registers it as a mobile web app
+    -   The initial HTML which begins loading the app and the requested form
 
--   An Enketo deployment is updated to a new version, and the user again accesses a form in offline-capable mode
--   When any change is made to the Service Worker script, configuration, or the version specified in `package.json`
+    This cache is cleared and rebuilt when an Enketo deployment is updated to a new version, and the user again accesses a form in offline-capable mode.
 
-Apart from those cases, onced cached these resources will always be retrieved from Cache Storage without performing a network request.
+-   Resources associated with individual forms:
 
-**Important note:** after such an update, previously cached forms must be requested in offline-capable mode _while online_ to be re-added to the cache.
+    -   Media attachments used by the form
+    -   [External secondary instances](https://getodk.github.io/xforms-spec/#secondary-instances---external) used by the form
 
-#### Form and form attachments
+    The resources in this cache will be updated when Enketo detects that they have changed.
+
+For production Enketo deployments, these resources will always be retrieved from the cache if present, without performing additional network requests other than to determine whether they have been updated. To aid maintenance and improvement of Enketo's offline functionality, the requests _are_ performed in development mode.
```

Comment:
Help me get the flow in more detail here. I'm online, I go to an offline-capable form. There's no Enketo update but there are form updates. Here's what I believe happens and as I understand it this is unchanged by this PR:
- The form loads immediately from cache
- Enketo polls the form server and determines there's an update (either of the form definition or of anything in the manifest or both)
- Enketo shows the banner prompting me to refresh
- When I refresh, Enketo requests the updated form and form attachments. It requests all of them even if only one of them has changed. E.g if only the form has changed, attachments are still updated. If only a 200kb attachment has been updated, it still downloads that and the 60MB video that wasn't updated.

Why is there a delay in the version change polling? Is that something Enketo-specific or just the APIs?

Reply:
Actually, I don't think that's what happens anymore. Here's now what I'm thinking is going on. I think this "similar but additive change was made to check for form updates on page load as well."

Before:
- The form loads immediately from cache
- After a short delay, Enketo polls the form server and determines there's an update (either of the form definition or of anything in the manifest or both)
- Enketo shows the banner prompting me to refresh
- When I refresh, Enketo requests the updated form and form attachments. It requests all of them even if only one of them has changed. E.g if only the form has changed, attachments are still updated. If only a 200kb attachment has been updated, it still downloads that and the 60MB video that wasn't updated.

After:
- The form does not load immediately
- Enketo determines there's an update (either of the form definition or of anything in the manifest or both)
- Enketo requests the updated form and form attachments. It requests all of them even if only one of them has changed. E.g if only the form has changed, attachments are still updated. If only a 200kb attachment has been updated, it still downloads that and the 60MB video that wasn't updated.
- The form loads

Reply:
For form updates, the flow you describe here is almost right. When a change is detected, the form cache is fully updated in the background before the banner is shown. Once prompted, if the user reloads they'll again request the form from the cache. A substantive change in this PR is that the initial check is performed immediately without a delay. (Currently this is blocking, but that was a mistake I'll correct shortly.)

Likewise, a substantive change in this PR is that the initial check for service worker updates is no longer delayed (and also non-blocking). There's an inherent delay for service worker updates because once one has been registered it's active immediately on load—i.e. before an updated one can be registered, installed and activated.

I think we should also maybe limit the initial check auto-reload to a shorter time span (e.g. 500ms-1s?), in case for some reason the service worker update takes longer and would potentially interrupt the user's work. (But maybe auto-save adequately addresses this?)

Reply:
> the form cache is fully updated in the background before the banner is shown

Nice!

> this is blocking, but that was a mistake I'll correct shortly

Ok, great. I'm not entirely sure which is best but seems reasonable to leave close to the current behavior. The improvement is that the banner will show up right away if on a fast connection. The pause is disconcerting, I find.

> I think we should also maybe limit the initial check auto-reload to a shorter time span (e.g. 500ms-1s?

What is the user experience of the auto-reload? Does it feel like a visible page refresh? There's currently no limit, right? Going to a banner experience after a second feels like a reasonable approach.

Reply:
> Does it feel like a visible page refresh?

Yes and it's weird.

Follow-on work would be to move the service worker registration earlier in the flow to avoid the double reload.